apigee/analytics/analytics.go

```diff
@@ -16,96 +16,386 @@ package analytics
+	// AnalyticsURL is where analytics get uploaded to.
+	AnalyticsURL string
+}
+
+// NewManager constructs and starts a new Manager. Call Close when you are done.
+func NewManager(env adapter.Env, opts Options) (*Manager, error) {
+	m, err := newManager(opts)
+	if err != nil {
+		return nil, err
 	}
-	if auth.Organization() == "" || auth.Environment() == "" {
-		return fmt.Errorf("organization and environment are required in auth: %v", auth)
+	m.Start(env)
+	return m, nil
+}
+
+func newManager(opts Options) (*Manager, error) {
+	if opts.BufferPath == "" {
+		opts.BufferPath = defaultBufferPath
+	}
+	if opts.AnalyticsURL == "" {
+		opts.AnalyticsURL = defaultAnalyticsURL
 	}
 
-	for i := range records {
-		records[i].RecordType = axRecordType
+	if _, err := os.Stat(opts.BufferPath); err != nil {
+		if os.IsNotExist(err) {
+			// Attempt to create it.
+			if err := os.MkdirAll(opts.BufferPath, os.ModeDir); err != nil {
+				return nil, err
+			}
+		} else {
+			return nil, fmt.Errorf("stat buffer path: %s", err)
+		}
+	}
 
-		// populate from auth context
-		records[i].DeveloperEmail = auth.DeveloperEmail
-		records[i].DeveloperApp = auth.Application
-		records[i].AccessToken = auth.AccessToken
-		records[i].ClientID = auth.ClientID
+	return &Manager{
+		close: make(chan bool),
+		client: &http.Client{
+			Timeout: httpTimeout,
+		},
+		now:                time.Now,
+		collectionInterval: defaultCollectionInterval,
+		bufferPath:         opts.BufferPath,
+		analyticsURL:       opts.AnalyticsURL,
+		instanceID:         uuid.New().String(),
+	}, nil
+}
 
-		// todo: select best APIProduct based on path, otherwise arbitrary
-		if len(auth.APIProducts) > 0 {
-			records[i].APIProduct = auth.APIProducts[0]
+// Start starts the manager.
+func (m *Manager) Start(env adapter.Env) {
+	m.log = env.Logger()
+	env.ScheduleDaemon(func() {
+		m.uploadLoop()
+	})
+}
+
+// Close shuts down the manager.
+func (m *Manager) Close() {
+	if m == nil {
+		return
+	}
+	m.close <- true
+	if err := m.upload(); err != nil {
+		m.log.Errorf("Error pushing analytics: %s", err)
+	}
+}
+
+// uploadLoop runs a timer that periodically pushes everything in the buffer
+// directory to the server.
+func (m *Manager) uploadLoop() {
+	t := time.NewTimer(m.collectionInterval)
+	for {
+		select {
+		case <-t.C:
+			if err := m.upload(); err != nil {
+				m.log.Errorf("Error pushing analytics: %s", err)
+			}
+		case <-m.close:
+			t.Stop()
+			return
 		}
 	}
+}
 
-	axURL := auth.ApigeeBase()
-	axURL.Path = path.Join(axURL.Path, fmt.Sprintf(axPath, auth.Organization(), auth.Environment()))
+// upload sends any buffered data to UAP.
+func (m *Manager) upload() error {
+	files, err := ioutil.ReadDir(m.bufferPath)
+	if err != nil {
+		return fmt.Errorf("ReadDir(%s): %s", m.bufferPath, err)
+	}
 
-	request := request{
-		Organization: auth.Organization(),
-		Environment:  auth.Environment(),
-		Records:      records,
+	var errOut error
+	var success int
+	for _, fi := range files {
```

Comment:
Do we need to consider a pool of upload threads?

Reply:
Not sure, do we? Do we expect to get SendRecords() calls faster than we can upload these files?

Reply:
Well, don't worry about it for now. We'll just have to keep an eye on it.